pkg/admission/server.go

```diff
@@ -0,0 +1,161 @@
+	"k8s.io/klog/v2"
+	"net/http"
+	"path/filepath"
+	"reflect"
+	v1alpha1 "sigs.k8s.io/service-apis/apis/v1alpha1"
+)
+
+const (
+	tlsDir      = `/run/secrets/tls`
+	tlsCertFile = `tls.crt`
+	tlsKeyFile  = `tls.key`
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func log500(w http.ResponseWriter, err error) {
+	klog.Errorf("failed to read request from client: %v\n", err)
+	http.Error(w, err.Error(), http.StatusInternalServerError)
+	return
+}
+
+// ServeHTTP parses AdmissionReview requests and responds back
+// with the validation result of the entity.
+func ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Body == nil {
+		klog.Errorf("received request with empty body")
+		http.Error(w, "admission review object is missing",
+			http.StatusBadRequest)
+		return
+	}
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log500(w,err)
+	}
+	review := admission.AdmissionReview{}
+	if err := json.Unmarshal(data, &review); err != nil {
+		klog.Errorf("failed to parse AdmissionReview object: %v\n", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	response, err := handleValidation(*review.Request)
+	if err != nil {
+		log500(w,err)
+	}
+	review.Response = response
+	data, err = json.Marshal(review)
+	if err != nil {
+		log500(w,err)
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		log500(w,err)
+	}
+}
+
+var (
+	httpRoute = meta.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  v1alpha1.SchemeGroupVersion.Version,
+		Resource: "httproutes",
+	}
+	gateway = meta.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  v1alpha1.SchemeGroupVersion.Version,
+		Resource: "gateway",
+	}
+)
+
+func routesEqual(new v1alpha1.HTTPRoute, old v1alpha1.HTTPRoute) bool {
+	if reflect.DeepEqual(new, old) {
+		return true
+	}
+	return false
+}
+
+func handleValidation(request admission.AdmissionRequest) (
+	*admission.AdmissionResponse, error) {
+	var response admission.AdmissionResponse
+	var ok bool
+	var message string
+	var err error
+
+	if request.Resource == httpRoute {
+		hRoute := v1alpha1.HTTPRoute{}
+		deserializer := codecs.UniversalDeserializer()
+		_, _, err = deserializer.Decode(request.Object.Raw, nil, &hRoute)
+		if err != nil {
+			return nil, err
+		}
+		switch request.Operation {
+		case admission.Create:
+			ok, message, err = ValidateHTTPRoute(hRoute)
+			if err != nil {
+				return nil, err
+			}
+		case admission.Update:
+			oldRoute := v1alpha1.HTTPRoute{}
+			_,_,err = deserializer.Decode(request.OldObject.Raw, nil, &oldRoute)
+			if err != nil {
+				return nil, err
+			}
+			// validate if routes are changed
+			if !routesEqual(hRoute, oldRoute) {
+				ok, message, err = ValidateHTTPRoute(hRoute)
+				if err != nil {
+					return nil, err
+				} else {
+					ok = true
+				}
+			}
+		default:
+			return nil, fmt.Errorf("unknown operation '%v'", string(request.Operation))
+		}
```

Comment:
We shouldn't differentiate between Create and Update, at least not at this point to keep the code simple.